### PR TITLE
[MPS] Add support for bf16 autocast

### LINKED ISSUE #139386: [MPS] Extend autocast support to bf16
### 🚀 The feature, motivation and pitch

In #99272, autocast support was added for fp16 (`torch.float16`) in MPS. We would like to extend this functionality to include bf16 (`torch.bfloat16`). Most of the code required for this feature can be referenced in #99272, so please review prior commits for guidance.

### Key Requirements
- **macOS Support**: bf16 support is only available on macOS 14 and later; the feature should include checks to ensure compatibility.
- **Error Handling**: Implement clear error messages when bf16 is used on unsupported macOS versions.
- **Testing**: Add comprehensive unit tests for bf16 autocast, with emphasis on edge cases.
- **Documentation**: Update relevant documentation to include bf16 support details, including version requirements, usage examples, and potential performance considerations (if applicable).

### Alternatives

_No response_

### Additional context

_No response_

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

### PR DESCRIPTION
**Pull request by hvaara**
This PR adds support for bf16 autocast. Most of the code and ideas are copied from #99272.

Most of the heavy lifting was done by AI.

Fixes #139386

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: test/test_autocast.py [hunk 1]
```diff
@@ -350,11 +352,30 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
-            UserWarning, "MPS Autocast only supports dtype of torch.float16 currently."
+            UserWarning,
+            "MPS Autocast only supports dtype of torch.bfloat16 and torch.float16 currently.",
         ):
-            with torch.autocast(device_type="mps", dtype=torch.bfloat16):
+            with torch.autocast(device_type="mps", dtype=torch.float32):
                 _ = torch.ones(10)
 
+    # torch.bfloat16 is only supported on macOS 14 and above.
+    @expectedFailureMPSPre14
```
Comment (hvaara) [inline test/test_autocast.py:362]: Is it okay to use the xfail predicated on version here?

cc @malfet

Comment (malfet) [reply test/test_autocast.py:362]: If it works, than sounds good to me :)

File: torch/amp/autocast_mode.py [hunk 2]
```diff
@@ -323,14 +324,23 @@ def __init__(
...
-                    "MPS Autocast only supports dtype of torch.float16 currently."
+                error_message = (
+                    "In MPS autocast, but the target dtype is not supported. Disabling autocast.\n"
+                    "MPS Autocast only supports dtype of torch.bfloat16 and torch.float16 currently."
                 )
                 warnings.warn(error_message)
                 enabled = False
+            elif self.fast_dtype == torch.bfloat16:
+                mac_ver = platform.mac_ver()[0]
+                if mac_ver and tuple(map(int, mac_ver.split("."))) < (14, 0):
```
Comment (malfet) [inline torch/amp/autocast_mode.py:337]: This function actually checks if MPS framework supports certain functions
```suggestion
                if not torch.backends.mps.is_macos_or_newer(14, 0):
```

File: test/test_autocast.py [hunk 3]
```diff
@@ -350,11 +352,30 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            self.assertEqual(result.dtype, torch.bfloat16)
+
+    @patch("platform.mac_ver", return_value=("13.0", ("", "", ""), ""))
+    def test_mps_autocast_bfloat16_unsupported(self, mock_mac_ver):
+        with self.assertWarnsRegex(
+            UserWarning,
+            "In MPS autocast, but the target dtype torch.bfloat16 is not supported on macOS versions below 14. Disabling autocast.",
+        ):
+            with torch.amp.autocast(device_type="mps", dtype=torch.bfloat16):
+                _ = torch.ones(10, device="mps")
```
Comment (malfet) [inline test/test_autocast.py:377]: This test feels redundant, as previous test will already xfail on older MacsOS and pass on new one

Comment (hvaara) [reply test/test_autocast.py:377]: Yes, assuming we test on this platform version. I'll remove the test.

File: torch/amp/autocast_mode.py (hunk 2 above)
Comment (hvaara) [reply torch/amp/autocast_mode.py]: Good catch! :)

**Comment by hvaara** [comment]
@malfet is this good to go?

**Review by malfet — APPROVED** [review]
LGTM

**Comment by ghsanti** [comment]
this is great, is it possible to install it already for mac os ?

**Comment by hg0428** [comment]
What happened here?
What's the progress?
How do I use autocast with MPS?

**Comment by hvaara** [comment]
@hg0428 The PR was merged to `main` in https://github.com/pytorch/pytorch/commit/bc69a19139c33c8339b8ee8df3faa3f4aa765685, and is available in stable since `v2.6.0`. PyTorch's merge process is handled oob which is why the PR is closed on GitHub.

For a primer on how to use autocast please see https://pytorch.org/docs/stable/amp.html and https://pytorch.org/docs/stable/notes/amp_examples.html. If it's still not clear, please ask on https://discuss.pytorch.org/.

**Pull request closed without merging**